lib/Dialect/Seq/Transforms/LowerSeqFIFO.cpp

```diff
@@ -0,0 +1,197 @@
+//===- LowerSeqFIFO.cpp - seq.fifo lowering -------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/SV/SVOps.h"
+#include "circt/Dialect/Seq/SeqOps.h"
+#include "circt/Dialect/Seq/SeqPasses.h"
+#include "circt/Support/BackedgeBuilder.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace circt;
+using namespace seq;
+
+namespace {
+
+struct FIFOLowering : public OpConversionPattern<seq::FIFOOp> {
+public:
+  using OpConversionPattern::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(seq::FIFOOp mem, OpAdaptor adaptor,
```

Comment:
One thing we've talked about is having a circt.seq verilog package and modules as a stand alone library to eliminate open-coded complex implementations (memories also).  Essentially if you can write the parametric module by hand and it's non-trivial, we should probably just be doing that in verilog source.

Reply:
We could have a generic pass which looks for ops with a special keyword to indicate the name of the module to instantiate which implements a particular op.

Reply:
+1 for this - essentially a generic "externalize" pass (which is already being done for e.g. clock gates).